### PR TITLE
Refactor Encrypted chat to component

### PR DESCRIPTION
**Pull request by KoalaSat**
## What does this PR do?
A refactor of the `EncryptedChat` component to help with the implementation of the chat turtle mode.

No behavior changes should happen

## Checklist before merging
- [ x ] If it's a frontend feature, I have ran prettier `cd frontend; npm run format`. If it's a mobile app feature I ran `cd mobile; npm run format`.

### CODE REVIEW

**Review by Reckless-Satoshi — COMMENTED** [review]
Hey @KoalaSat `EncryptedChat` has never looked soo elegant as in this PR :smile:

I left a few comments with my findings.

File: frontend/src/components/TradeBox/EncryptedChat/index.tsx [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+  const scrollToBottom = () => {
+    messageEndRef.current?.scrollIntoView({ behavior: 'smooth' });
+  };
+
+  const onMessage: (message: any) => void = (message) => {
+    const dataFromServer = JSON.parse(message.data);
+    console.log('Got reply!', dataFromServer.type);
+    console.log('PGP message index', dataFromServer.index, ' latestIndex ', latestIndex);
+    if (dataFromServer) {
+      console.log(dataFromServer);
```
[the remark this answers is not in the document]
Comment (KoalaSat) [reply frontend/src/components/TradeBox/EncryptedChat/index.tsx:122]: this one took me a while to figure out, looks like not always the cookies (`pub_key`, `enc_priv_key` and `robot_token`) are generated. That's why the chat couldn't manage to connect, it seems to be something related to the test user you created. Looks like sometimes the cookies are created and others don't. I briefly tested with other users but now looks like https://robotest.onion.moe/ is down. I'll text it more in detail once it's up again

[the remark this answers is not in the document]
Comment (Reckless-Satoshi) [reply frontend/src/components/TradeBox/EncryptedChat/index.tsx:122]: Interesting. I believe the reconnect loop I see is not due to the cookies.. I used the same test users in robotest using the frontend in main and the chat worked as expected. On the other hand, the error I had happened on my local development node with freshly baked robots ^^

File: frontend/src/components/TradeBox/EncryptedChat/index.tsx [hunk 2]
```diff
@@ -47,10 +47,10 @@ const EncryptedChat: React.FC<Props> = ({ orderId, userNick }: Props): JSX.Eleme
   const messageEndRef = useRef<HTMLInputElement>(null);
   const [connected, setConnected] = useState<boolean>(false);
   const [peerConnected, setPeerConnected] = useState<boolean>(false);
-  const [ownPubKey, setOwnPubKey] = useState<string>();
-  const [ownEncPrivKey, setOwnEncPrivKey] = useState<string>();
+  const [ownPubKey] = useState<string>(systemClient.getCookie('pub_key') || '');
+  const [ownEncPrivKey] = useState<string>(systemClient.getCookie('enc_priv_key') || '');
```
Comment (Reckless-Satoshi) [inline frontend/src/components/TradeBox/EncryptedChat/index.tsx:51]: Aren't these keys malformed? OpenPGP wants "Armored ACII" with linebreaks, but cookies can only be a string. So if I remember well, `\n` is replaced with `\\` when saving to cookies. Maybe these two lines also need the `.split('\\').join('\n')` ?

File: frontend/src/components/TradeBox/EncryptedChat/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Reckless-Satoshi) [reply frontend/src/components/TradeBox/EncryptedChat/index.tsx:122]: If you have access to `Docker` I would strongly recommend the new way to develop Web frontend features that uses directly the onion sites and exposes them locally (so you can use any browser, no Requestly needed). 

https://github.com/Reckless-Satoshi/robosats/blob/main/setup.md#with-docker That way we do not rely anymore on "onion.moe". It uses by default the mainnet, but can be changed to robotest by editing `nodeapp/docker-compose.yml`

[the remark this answers is not in the document]
Comment (KoalaSat) [reply frontend/src/components/TradeBox/EncryptedChat/index.tsx:122]: all right I'll try it, thanks for all the hits!

**Comment by KoalaSat** [comment]
Never thought this apparently simple refactor would have been so challenging, these changes requires heavy testing, I'll leave some of the points I had to fix during the development to have an eye on it:

- Performance and memory usage
- List won't scrolling with new message
- Notification sound not being played
- Notification sound being displayed twice. This still happens from time to time to me on Firefox, but debugging only fires the compiler once, I believe it's related to the fact that this is a react hook component inside a non hook component (`TradeBox/index.js`), couldn't find a way to properly reproduce it.
- Taker appearing online when it was not

Other than that everything now looks like working!

**Comment by Reckless-Satoshi** [comment]
Hey Koala, thanks a lot for the effort with this component!

I have not tested deeply either, saw a few style bugs (message of texts printing too far right, etc) and took the change to re-do the style. I think we can merge as is, and test in Testnet for a few more days before moving to merge.

Please accept a small compensation of 500K Sats for the nightmare refactor of the chat component.

**Comment by KoalaSat** [comment]
As always, as been fun anyways 😄

**Comment by KoalaSat** [comment]
```
lnbc5m1p34alh5pp568vn3za7qvl4e7q38rxtaeckmp23uxau6tpl3y29aprdk8u08z9qdqu2askcmr9wssx7e3q2dshgmmndp5scqzpgxqyz5vqsp5nmsjhkana22xnjuau4ppvxcp7rygpy0hzhhljz5nyj8urhh0drrs9qyyssqf0wqju6nnu02fsjkeflcrjyd2ne4vw8gmgu4e5sgual90zew8v9xc6dt3y37a8vyem8wtxqdtvc6sj2ypknd7rsug3wvn93g5pjtajsqmxkmu6
```

**Comment by Reckless-Satoshi** [comment]
> ```
> lnbc5m1p34alh5pp568vn3za7qvl4e7q38rxtaeckmp23uxau6tpl3y29aprdk8u08z9qdqu2askcmr9wssx7e3q2dshgmmndp5scqzpgxqyz5vqsp5nmsjhkana22xnjuau4ppvxcp7rygpy0hzhhljz5nyj8urhh0drrs9qyyssqf0wqju6nnu02fsjkeflcrjyd2ne4vw8gmgu4e5sgual90zew8v9xc6dt3y37a8vyem8wtxqdtvc6sj2ypknd7rsug3wvn93g5pjtajsqmxkmu6
> ```

6581045d252d0fb9997c82faa6987b02eb4e9c85121a6e30e207d2ce742d4e32

**Pull request merged**